File: code/KernelEP.NET/lib/Infer.NET 2.6/Samples/F#/TestFSharp/TruncatedGaussian.fs
#light

namespace TruncatedGaussianTutorial

open System
open MicrosoftResearch.Infer
open MicrosoftResearch.Infer.Models
open MicrosoftResearch.Infer.Distributions
open MicrosoftResearch.Infer.Factors
open MicrosoftResearch.Infer.Maths
open MicrosoftResearch.Infer.FSharp

//-----------------------------------------------------------------------------------
// Infer.NET: F# script for a truncated Gaussian with a variable threshold
//-----------------------------------------------------------------------------------

// The Model
module truncated =
    let truncatedTestFunc() = 
        Console.WriteLine("\n====================Running Truncated Gaussian Tutorial======================\n");
        let threshold = (Variable.New<float>()).Named("threshold")
        let x = Variable.GaussianFromMeanAndVariance(0.0,1.0).Named("x")
        do Variable.ConstrainTrue( (x >> threshold))

        // The inference, looping over different thresholds
        let ie = InferenceEngine()
        ie.ShowProgress <- false
        threshold.ObservedValue <- -0.1

        for i = 0 to 10 do
          threshold.ObservedValue <- threshold.ObservedValue + 0.1  
          printfn "Dist over x given thresh of %A = %A" threshold.ObservedValue (ie.Infer<Gaussian>(x))
        

File: code/KernelEP.NET/lib/Infer.NET 2.6/Samples/C#/ExamplesBrowser/TruncatedGaussian.cs
using System;
using System.Collections.Generic;
using System.Text;
using MicrosoftResearch.Infer.Models;

namespace MicrosoftResearch.Infer.Tutorials
{
	[Example("Tutorials", "Inefficient implementation of a truncated Gaussian distribution.",Prefix="2a.")]
	public class TruncatedGaussian
	{
		public void Run()
		{
			for (double thresh = 0; thresh <= 1; thresh += 0.1)
			{
				Variable<double> x = Variable.GaussianFromMeanAndVariance(0, 1).Named("x");
				Variable.ConstrainTrue(x > thresh);
				InferenceEngine engine = new InferenceEngine();
				if (engine.Algorithm is ExpectationPropagation)
					Console.WriteLine("Dist over x given thresh of " + thresh + "=" + engine.Infer(x));
				else
					Console.WriteLine("This example only runs with Expectation Propagation");
			}
		}

	}


}
